File: apps/visualizer/backend/api/schemas/perspectives.py
"""Perspectives API response models — single source of truth for perspective shapes."""

from __future__ import annotations

from pydantic import BaseModel, ConfigDict, RootModel


class PerspectiveSummary(BaseModel):
    model_config = ConfigDict(extra='forbid')

    id: int
    slug: str
    display_name: str
    description: str
    active: bool
    optional: bool
    source_filename: str | None = None
    updated_at: str | None = None


class PerspectiveDetail(PerspectiveSummary):
    prompt_markdown: str
    created_at: str | None = None


class PerspectiveScore(BaseModel):
    """Nested perspective line item in image descriptions."""

    model_config = ConfigDict(extra='forbid')

    analysis: str
    score: int


class PerspectiveListResponse(RootModel[list[PerspectiveSummary]]):
    """``GET /api/perspectives/`` response body."""


def validate_perspective_summary(row: dict) -> dict:
    """Validate a list-row dict; raises on shape drift."""
    return PerspectiveSummary.model_validate(row).model_dump(mode='json')


def validate_perspective_detail(row: dict) -> dict:
    """Validate a detail-row dict; raises on shape drift."""
    return PerspectiveDetail.model_validate(row).model_dump(mode='json')

File: apps/visualizer/backend/src/api/schemas/perspectives.ts
/**
 * Perspectives API response models.
 *
 * `optional` is read-only over HTTP — it is derived from the markdown marker on
 * write (ADR-0012), never accepted from a request body.
 */
import { z } from '@hono/zod-openapi';

export const PerspectiveSummary = z
  .object({
    id: z.int(),
    slug: z.string(),
    display_name: z.string(),
    description: z.string(),
    active: z.boolean(),
    optional: z.boolean(),
    source_filename: z.string().nullish(),
    updated_at: z.string().nullish(),
  })
  .strict()
  .openapi('PerspectiveSummary');

export const PerspectiveDetail = PerspectiveSummary.extend({
  prompt_markdown: z.string(),
  created_at: z.string().nullish(),
})
  .strict()
  .openapi('PerspectiveDetail');

/** Nested perspective line item in image descriptions. */
export const PerspectiveScore = z
  .object({
    analysis: z.string(),
    score: z.int(),
  })
  .strict()
  .openapi('PerspectiveScore');

/** `GET /api/perspectives/` response body — a bare array. */
export const PerspectiveListResponse = z
  .array(PerspectiveSummary)
  .openapi('PerspectiveListResponse');

export type PerspectiveSummary = z.infer<typeof PerspectiveSummary>;
export type PerspectiveDetail = z.infer<typeof PerspectiveDetail>;
